### PR TITLE
Bugfix/routing segfault 777

### PR DESCRIPTION
**Pull request by yao-cqc**
This should address the bug raised in this https://github.com/CQCL/tket/issues/777#issuecomment-1508934186

Initial and final maps can be considered as mapping logical qubits to physical qubits. When adding an ancilla node `node[x]` during routing, the current logic adds the entry {`node[x]`, `node[x]`} to the initial and final maps. So we are essentially giving both the ancilla logical qubit and its physical mapping the same name `node[x]`. However, this is error prone since there might already exits a logical qubit with the name `node[x]`. For example, route a circuit twice.

This PR fixes this issue by naming the ancilla logical qubit `tk_ROUTING_ANCILLA_REG[i]`, and making sure this qubit was not previously present.

### CODE REVIEW

File: tket/src/Mapping/MappingFrontier.cpp [hunk 1]
```diff
@@ -763,22 +788,39 @@ void MappingFrontier::add_bridge(
...
+
+  this->bimaps_->initial.insert({placeholder_q, ancilla});
+  this->bimaps_->final.insert({placeholder_q, ancilla});
   this->ancilla_nodes_.insert(Node(ancilla));
-  UnitID uid_ancilla(ancilla);
-
-  unit_map_t update_map;
-  update_map.insert({uid_ancilla, uid_ancilla});
-
-  update_maps(this->bimaps_, update_map, update_map);
```
Comment (yao-cqc) [inline tket/src/Mapping/MappingFrontier.cpp:781]: I'm not sure if this actually does anything.

File: pytket/docs/changelog.rst [hunk 2]
```diff
@@ -24,6 +24,8 @@ Fixes:
 
 * Fix bug in `get_operator_expectation_value()` computation when operator
   includes `Pauli.I` terms.
+* Fix bug in routing when the circuit contains empty qubit wires and some of its qubits
```
Comment (sjdilkes) [inline pytket/docs/changelog.rst:27]: *Fix bug in routing code occurring in Circuits with qubit wires with no operations and some (other or same) qubits pre-labelled as Node from the Architecture being routed to.

File: tket/src/Mapping/MappingFrontier.cpp (hunk 1 above)
Comment (sjdilkes) [reply tket/src/Mapping/MappingFrontier.cpp:781]: I think if you're directly modifying this->bimaps_->initial/final yourself then we can probably avoid this? Also if tests are passing then seems fine to remove.

**Review by sjdilkes — CHANGES_REQUESTED** [review]
Just the minor edit to the changelog I think, code changes look good. 

Side note - have you tried running the prop tests a bunch? I appreciate the CI should capture any bugs but I found when dealing with this sort of bug originally that the proptests were helpful for occasionally generating a rogue Circuit. Not a very scientific approach, but I'd recommend quickly rerunning the proptests a bunch of times locally just to confirm nothing comes up.

**Comment by yao-cqc** [comment]
> Just the minor edit to the changelog I think, code changes look good.
> 
> Side note - have you tried running the prop tests a bunch? I appreciate the CI should capture any bugs but I found when dealing with this sort of bug originally that the proptests were helpful for occasionally generating a rogue Circuit. Not a very scientific approach, but I'd recommend quickly rerunning the proptests a bunch of times locally just to confirm nothing comes up.

Yes, I have run the prop test in a for loop, but nothing has come up.

**Review by sjdilkes — APPROVED** [review]

**Pull request merged**